tests/HTMLPurifier/Lexer/Html5LexTest.php

```diff
@@ -0,0 +1,748 @@
+                new HTMLPurifier_Token_Empty('br'),
+            )
+        );
+    }
+
+    /**
+     * Tokenize a scripts CDATA content.
+     *
+     * @return void
+     */
+    public function test_tokenizeHTML_scriptCDATAContents()
+    {
+        $this->config->set('HTML.Trusted', true);
+
+        $this->assertTokenization(
+            'Foo: <script>alert("<foo>");</script>',
+            array(
+                new HTMLPurifier_Token_Text('Foo: '),
+                new HTMLPurifier_Token_Start('script'),
+                new HTMLPurifier_Token_Text('alert("<foo>");'),
+                new HTMLPurifier_Token_End('script'),
+            )
+        );
+    }
+
+    /**
+     * Tokenize HTML entities in a comment.
+     *
+     * @return void
+     */
+    public function test_tokenizeHTML_entitiesInComment()
+    {
+        $this->assertTokenization(
+            '<!-- This comment < &lt; & -->',
+            array(new HTMLPurifier_Token_Comment(' This comment < &lt; & '))
+        );
+    }
+
+    /**
+     * Tokenize special characters inside an attribute.
+     *
+     * @return void
+     */
+    public function test_tokenizeHTML_attributeWithSpecialCharacters()
+    {
+        $this->assertTokenization(
+            '<a href="><>">',
+            array(new HTMLPurifier_Token_Empty('a', array('href' => '><>')))
+        );
+    }
+
+    /**
+     * Tokenize a self closing tag with a forward slash in an attribute value.
+     *
+     * @return void
+     */
+    public function test_tokenizeHTML_emptyTagWithSlashInAttribute()
+    {
+        $this->assertTokenization(
+            '<param name="src" value="http://example.com/video.wmv" />',
+            array(new HTMLPurifier_Token_Empty('param', array('name' => 'src', 'value' => 'http://example.com/video.wmv')))
+        );
+    }
+
+    /**
+     * Tokenize a style tag.
+     *
+     * @return void
+     */
+    public function test_tokenizeHTML_style()
+    {
+        $expected = array(
+            new HTMLPurifier_Token_Start('style', array('type' => 'text/css')),
+            new HTMLPurifier_Token_Text("\ndiv {}\n"),
+            new HTMLPurifier_Token_End('style'),
+        );
+
+        if (!defined('LIBXML_VERSION')) {
+            // LIBXML_VERSION is missing in early versions of PHP
+            // prior to 1.30 of php-src/ext/libxml/libxml.c (version-wise,
+            // this translates to 5.0.x. In such cases, punt the test entirely.
+            return;
+        } elseif (LIBXML_VERSION < 20628) {
+            // libxml's behavior is wrong prior to this version, so make
+            // appropriate accommodations
+            $expected = array(
+                new HTMLPurifier_Token_Start('style', array('type' => 'text/css')),
+                new HTMLPurifier_Token_Comment("\ndiv {}\n"),
+                new HTMLPurifier_Token_End('style'),
+            );
+        }
+
+        $this->assertTokenization(
+            '<style type="text/css"><!--
+div {}
+--></style>',
```

Comment:
This seems to differ to DOMLex... the content of the `<style>` on `DOMDocument` is of `nodeType` `4` (DOMCdataSection) but on `masterminds/html5` it's `DOMText`

htmlpurifier has some special handling that removes the `<!-- -->`
https://github.com/ezyang/htmlpurifier/blob/ab2887e423936c5f2f422c4462bc4ad1f382f712/library/HTMLPurifier/Lexer/DOMLex.php#L194

because the `nodeType` isn't what's expected, that's not running and the comments aren't being removed.

Not sure if there's any implications of this...

Reply:
Logged https://github.com/Masterminds/html5-php/issues/175

Reply:
Updated the test